aten/src/ATen/native/Cross.cpp

```diff
@@ -1,52 +1,71 @@
-  return out;
+  auto dim = _default_cross_dim(dimension, input.sizes());
+  return at::linalg_cross(input, other, dim);
 }
 
 Tensor & cross_out(const Tensor & input, const Tensor & other, const c10::optional<int64_t> dimension, Tensor & out) {
-  auto device_res = input.device().type();
-  TORCH_CHECK(device_res == kCPU || device_res == kCUDA, "cross only supports CPU and CUDA devices, out got: ", device_res);
+  auto dim = _default_cross_dim(dimension, input.sizes());
+  return at::linalg_cross_out(out, input, other, dim);
+}
+
+Tensor linalg_cross(const Tensor & input, const Tensor & other, const int64_t dimension) {
+  Tensor out = at::empty({0}, input.options());
+  native::linalg_cross_out(input, other, dimension, out);
+  return out;
+}
+
+Tensor & linalg_cross_out(const Tensor & input, const Tensor & other, const int64_t dimension, Tensor & out) {
   auto device1 = input.device().type();
-  TORCH_CHECK(device1 == kCPU || device1 == kCUDA, "cross only supports CPU and CUDA devices, input got: ", device1);
-  auto device2 = other.device().type();
-  TORCH_CHECK(device2 == kCPU || device2 == kCUDA, "cross only supports CPU and CUDA devices, other got: ", device2);
-  TORCH_CHECK(device_res == device1, "out and input must have the same device type. out: ", device_res, " input: ", device1);
-  TORCH_CHECK(device1 == device2, "input and other must have the same device type. input: ", device1, " other: ", device2);
-  TORCH_CHECK(!out.is_cuda() || out.get_device() == input.get_device(), "device of out (", input.get_device(), ") must match device of input (", other.get_device(), ")");
-  TORCH_CHECK(!input.is_cuda() || input.get_device() == other.get_device(), "device of input (", input.get_device(), ") must match device of other (", other.get_device(), ")");
-  TORCH_CHECK(input.dim() == other.dim(), "inconsistent tensors dimensions input: ", input.dim(), " other: ", other.dim());
-  TORCH_CHECK(input.sizes() == other.sizes(), "inconsistent tensors sizes input: ", input.sizes(), " other: ", other.sizes());
-
-  int64_t dim = -1;
-  if(!dimension.has_value()) {
-    for(int64_t i = 0; i < input.dim(); i++) {
-      if(input.size(i) == 3) {
-        dim = i;
-        break;
-      }
-    }
-    TORCH_CHECK(dim >= 0, "no dimension of size 3 in input");
-  } else {
-    dim = maybe_wrap_dim(dimension.value(), input.dim());
-    TORCH_CHECK(input.size(dim) == 3, "dimension ", dimension.value(), " does not have size 3");
-  }
 
-  if (out.sizes() != input.sizes()) {
-    out.resize_as_(input);
+  // Check for possible broadcasting
+  auto input_size = input.sizes();
+  auto other_size = other.sizes();
+  auto out_size = infer_size(input_size, other_size);
+
+  Tensor input_broadcasted = input;
+  Tensor other_broadcasted = other;
+
+  if (not input_size.equals(out_size)){
+    input_broadcasted = input.expand(out_size);
+  }
+  if (not other_size.equals(out_size)){
+    other_broadcasted = other.expand(out_size);
```

Comment:
why not simply `input_broadcasted = input.expand(out_size);`?

Reply:
I thought with the `if` condition we can skip broadcasting in case it is not required.

Reply:
I don't think this optimisation is needed here. Readability is more important.

If you want to go the extra mile, you can have a look at the functions in `ExpandUtils.h` that use `MaybeOwned<Tensor>`s, but I think that it's not really necessary here. Up to you.

Reply:
Sure! I'll just simplify that for better readability.